Add blog about CHERI in safety-critical systems. (#22)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2024-11-25-cheri-myths-safety-critical.markdown

---
layout: post
title:  "CHERI Myths: CHERI is incompatible with safety-critical systems"
date:   2024-11-25
categories: cheri myths
author: David Chisnall
---

A few times, especially in the context of the Digital Catapult Technology Access Programmes, I've heard people claim that the fact that CHERI traps on memory-safety errors means that it can't be used in safety-critical systems.
I understand some of the rationale behind these claims, but the claims themselves seem the result of misunderstanding the current baseline.
The claims are usually of the form 'our code may contain memory-safety bugs but they don't impact the safety-critical parts and so it's better to leave them'.
This is a dangerous view, as I will explain below.

> ## Why are memory-safety errors bad?

Reviewer: ```suggestion
## Why are memory-safety errors so bad?
```
Maybe add 'in safety-critical systems'?

Before we go into detail about what CHERI gives and why this may or may not be problematic, it's worth reconsidering why memory safety bugs are bad.
We often quote the Microsoft and Google numbers that 70% of vulnerabilities come from memory-safety bugs but if you dig deeper you'll see that the severity of memory-safety bugs is often disproportionately high.
The reason for this is the 'spooky action at a distance' problem.

If you have a typical logic error, you may go down the wrong branch in your code.
For example, some of the most famous early UNIX bugs involved the following pattern, which compilers will now warn about:

```c
if (uid = 0)
{
    // Root is allowed to do this
}
```

> The programmer thought that they were checking the user ID was the root user, but in fact they were assigning the root user ID to the `uid` variable.

Reviewer: ```suggestion
The programmer thought that they were checking the user ID was the root user, but in fact they were assigning the root user ID to the `uid` variable (because of the distinction between `=` and `==` in C-like languages).
```

This kind of bug is often a critical vulnerability because it can directly lead to privilege elevation.
You can tell, because you can look at where the `uid` variable is used later and follow the control flow throughout the source code of the program.

With a memory-safety bug, you cannot do this.

> If you write outside the bounds of a buffer, or through a dangling pointer, then you have stepped outside of the language's abstract machine.

Reviewer: ```suggestion
If you write outside the bounds of a buffer, or through a dangling pointer, then you have stepped outside of the language's *abstract machine* and into the world of *undefined behaviour*.
```

For the most part, MISRA C code should not trap with CHERI, and if it does then you probably haven't actually followed the MISRA C requirements.

The same is true of Rust.
If you're using purely safe Rust, your code will never trap on a CHERI system.
If you're using `unsafe` in small, locally scoped places, you can provide guarantees that are not visible in the source code to ensure the same properties.
If you write your entire program in unsafe Rust (no one does this!) then you should expect traps.

So do I need CHERI if I'm using MISRA C or Rust?
[I've answered that before](/cheri/myths/2024/08/28/cheri-myths-safe-languages.html), but broadly yes.
CHERI lets you enforce properties even in the unsafe dialects of safe languages, in C and assembly code linked with them, and on binary-only components or components compiled with an untrusted compiler (or code that may be written specifically to exploit bugs in a trusted compiler).

## What is your baseline?

To answer the complaint about CHERI trapping, it's worth considering what happens if you have a memory-safety bug today.
You will see one of the following:

>  - It's deterministically benign, reads some constant or write over something that's never read.

Reviewer: write -> writes

I also think 'benign' should be in scare quotes everywhere...

>  - It's deterministically out of a mapped / accessible region will be caught by your MMU / MPU and trap (so we are not worse)

Reviewer: - It's deterministically outside of a mapped / accessible region so will be caught by your MMU / MPU and trap (therefore CHERI is no worse)

## What does CHERIoT do with traps?

CHERIoT RTOS provides three ways for a compartment to handle CHERI failures (and other synchronous trap, such as invalid instruction errors):

 - If there is enough stack space, a compartment may provide an error handler that runs with a copy of the register file and can inspect and modify the state.
 - If there is not enough stack space or the first kind of error handler is not provided, the compartment may provide a *stackless* error handler.
   This is used to implement our structured error handling mechanisms.
 - If neither is provided, or the error handler returns the force-unwind result, the current compartment call is terminated and control returns to the caller with an error code in the return register.

For stateless compartments, the last is often the right approach: if something goes wrong, let the caller know.
This is often the best fit approach for safety-critical systems, where things should *never* go wrong and you've got static analysis receipts to demonstrate this.

> If you're relying on the notion of a 'benign' error, the first kind of error handler helps you emulate the current behaviour (though I don't recommend it if you actually care about correctness!).

Reviewer: ```suggestion
If you're relying on the notion of a 'benign' error, the first kind of error handler could help you emulate the current behaviour (though I don't recommend it if you actually care about correctness!).
```

You can decode the current instruction and determine whether it's a load or store.
If it's a store, you can skip it, if it's a load you can also put a zero in the target register.
Ideally, you'd also write some telemetry that would let you fix the bug.

> This turns your 'benign' memory-safety bugs into ones that have deterministic behaviour that is guaranteed not to impact the rest of the system.

Reviewer: ```suggestion
This turns your 'benign' memory-safety bugs into ones that have deterministic behaviour that is guaranteed not to impact the rest of the system, but your program's behaviour is still completely arbitrary.
Hopefully this example helps to demonstrate there really is no such thing as a 'benign' memory-safety error.
```
